libs/core/executors/include/hpx/executors/fork_join_executor.hpp

```diff
@@ -87,12 +89,38 @@ namespace hpx { namespace execution { namespace experimental {
                 hpx::concurrency::detail::contiguous_index_queue<std::uint32_t>;
             using queues_type =
                 std::vector<hpx::util::cache_aligned_data<queue_type>>;
-            using thread_states_type = std::vector<
-                hpx::util::cache_aligned_data<std::atomic<thread_state>>>;
-            using thread_function_helper_type = void(void*, void const*, void*,
-                std::size_t, std::size_t, loop_schedule, queues_type&,
-                thread_states_type&, hpx::lcos::local::spinlock&,
-                std::exception_ptr&);
+
+            struct region_data_type;
+            using thread_function_helper_type = void(region_data_type&,
+                std::size_t, std::size_t, queues_type&,
+                hpx::lcos::local::spinlock&, std::exception_ptr&) noexcept;
+
+            // Members that change for each parallel region.
+            struct region_data
+            {
+                // the thread state for each of the executed threads
+                std::atomic<thread_state> state_;
+
+                // The helper function that does the actual work for a single
+                // parallel region.
+                std::atomic<thread_function_helper_type*>
+                    thread_function_helper_;
+
+                // Pointers to inputs to bulk_sync_execute.
+                void* element_function_;
+                void const* shape_;
+                void* argument_pack_;
+            };
```

Comment:
Did you separate these out for performance or readability? I'm asking because of the potential for false sharing, but as far as I can tell this should be ok. A worker thread only changes its own `state_` at the end of a region.

Reply:
I wanted for those variables to always be on the same cache line. The core always loads the state, which makes those other values available right away. This change has improved the performance significantly. I will post a performance comparison plot once we're done with this.